cmd/createtree/main_test.go

```diff
@@ -43,147 +61,101 @@ func marshalAny(p proto.Message) *any.Any {
-	pkcs11Tree := *defaultTree
-	pkcs11Tree.PrivateKey, err = ptypes.MarshalAny(&keyspb.PKCS11Config{
-		TokenLabel: "log",
-		Pin:        "1234",
-		PublicKey: `-----BEGIN PUBLIC KEY-----
-MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQC7/tWwqUXZJaNfnpvnqiaeNMkn
-hKusCsyAidrHxvuL+t54XFCHJwsB3wIlQZ4mMwb8mC/KRYhCqECBEoCAf/b0m3j/
-ASuEPLyYOrz/aEs3wP02IZQLGmihmjMk7T/ouNCuX7y1fTjX3GeVQ06U/EePwZFC
-xToc6NWBri0N3VVsswIDAQAB
------END PUBLIC KEY-----
-`,
-	})
-	if err != nil {
-		t.Fatalf("MarshalAny(PKCS11Config): %v", err)
+func runTest(t *testing.T, tests []*testCase) {
+	keyHandlers["test"] = func() (proto.Message, error) {
+		return &empty.Empty{}, nil
 	}
 
-	emptyPKCS11Path := *validOpts
-	emptyPKCS11Path.privateKeyType = "PKCS11ConfigFile"
-
-	tests := []struct {
-		desc      string
-		opts      *createOpts
-		createErr error
-		wantErr   bool
-		wantTree  *trillian.Tree
-	}{
-		{desc: "validOpts", opts: validOpts, wantTree: defaultTree},
-		{desc: "nonDefaultOpts", opts: &nonDefaultOpts, wantTree: &nonDefaultTree},
-		{desc: "defaultOptsOnly", opts: newOptsFromFlags(), wantErr: true}, // No mandatory opts provided
-		{desc: "emptyAddr", opts: &emptyAddr, wantErr: true},
-		{desc: "invalidEnumOpts", opts: &invalidEnumOpts, wantErr: true},
-		{desc: "invalidKeyTypeOpts", opts: &invalidKeyTypeOpts, wantErr: true},
-		{desc: "emptyPEMPath", opts: &emptyPEMPath, wantErr: true},
-		{desc: "emptyPEMPass", opts: &emptyPEMPass, wantErr: true},
-		{desc: "PEMKeyFile", opts: &pemKeyOpts, wantErr: false, wantTree: &pemKeyTree},
-		{desc: "createErr", opts: validOpts, createErr: errors.New("create tree failed"), wantErr: true},
-		{desc: "PKCS11Config", opts: &pkcs11Opts, wantErr: false, wantTree: &pkcs11Tree},
-		{desc: "emptyPKCS11Path", opts: &emptyPKCS11Path, wantErr: true},
+	server, lis, stopFn, err := startFakeServer()
+	if err != nil {
+		t.Fatalf("Error starting fake server: %v", err)
 	}
+	defer stopFn()
 
 	ctx := context.Background()
 	for _, test := range tests {
 		server.err = test.createErr
 
-		tree, err := createTree(ctx, test.opts)
-		switch hasErr := err != nil; {
-		case hasErr != test.wantErr:
-			t.Errorf("%v: createTree() returned err = '%v', wantErr = %v", test.desc, err, test.wantErr)
-			continue
-		case hasErr:
-			continue
-		}
-
-		if diff := pretty.Compare(tree, test.wantTree); diff != "" {
-			t.Errorf("%v: post-createTree diff:\n%v", test.desc, diff)
-		}
+		// Wrap in a func so that defer can be used to cleanup after each iteration.
+		func() {
+			defer flagsaver.Save().Restore()
+			*adminServerAddr = lis.Addr().String()
+			*privateKeyFormat = "test"
+			if test.setFlags != nil {
+				test.setFlags()
+			}
+
+			tree, err := createTree(ctx)
+			switch hasErr := err != nil; {
+			case hasErr != test.wantErr:
+				t.Errorf("%v: createTree() returned err = '%v', wantErr = %v", test.desc, err, test.wantErr)
+				return
+			case hasErr:
+				return
+			}
+
+			if diff := pretty.Compare(tree, test.wantTree); diff != "" {
```

Comment:
Please use `proto.Equal()` when comparing protos, `pretty.Compare` doesn't know about internal proto fields, so it may diff in cases where protos are meant to be equal.

The suggested idiom is:

```go
if !proto.Equal(pb1, pb2) {
    diff := pretty.Compare(pb1, pb2)
    t.Errorf("%v: post-Foo diff (-got +want):\n%v", test.desc, diff)
}
```

(Yep, it's a bit verbose, I know :/ )